quantization/LowPrecisionQuantization.h

```diff
@@ -0,0 +1,222 @@
+
+    float min = std::numeric_limits<float>::max();
+    float max = std::numeric_limits<float>::min();
+
+    for (uint32_t dim_index = 0; dim_index < _dimension; dim_index++) {
+      auto diff = vector[dim_index] - _mean_vector[dim_index];
+      if (diff < min) {
+        min = diff;
+      }
+      if (diff > max) {
+        max = diff;
+      }
+    }
+    return {min, max};
+  }
+
+  /**
+   * @brief Quantizes a given vector following the locally-adaptive quantization
+   * scheme. quantized_vector must be pre-allocated.
+   *
+   * @param vector
+   * @param quantized_vector
+   */
+  template <typename T>
+  void quantizeVector(const float *vector, T *quantized_vector) const {
+    assert(_is_trained);
+    assert(sizeof(T) == _num_bits / 8);
+
+    auto [min, max] = getMinMax(vector);
+
+    float delta = (max - min) / ((1 << _num_bits) - 1);
+
+    for (uint32_t dim_index = 0; dim_index < _dimension; dim_index++) {
+      auto first_term = (vector[dim_index] - min) / delta;
+      first_term += 0.5;
+      quantized_vector[dim_index] =
+          static_cast<T>(delta * std::floor(first_term) + min);
+    }
+  }
+
+  template <typename data_t>
+  float distanceImpl(const void *x, const void *y,
+                     bool asymmetric = false) const {
+    bool metric_euclidean = _metric_type == METRIC_TYPE::EUCLIDEAN;
+
+    if (_num_bits == 16) {
+      void *x_ptr = (void *)x;
+
+      if (asymmetric) {
+        std::unique_ptr<int16_t[]> quantized_x =
+            std::make_unique<int16_t[]>(_dimension);
+        quantizeVector<int16_t>((float *)x, quantized_x.get());
+        x_ptr = (void *)quantized_x.get();
+      }
+      float distance =
+          metric_euclidean
+              ? std::get<0>(_distance).distanceImpl<int16_t>(x_ptr, y)
+              : std::get<1>(_distance).distanceImpl<int16_t>(x_ptr, y);
+      return distance;
+
+    } else if (_num_bits == 8) {
+      void *x_ptr = (void *)x;
+      if (asymmetric) {
+        std::unique_ptr<int8_t[]> quantized_x =
+            std::make_unique<int8_t[]>(_dimension);
+        quantizeVector<int8_t>((float *)x, quantized_x.get());
+        x_ptr = (void *)quantized_x.get();
+      }
+
+      float distance =
+          metric_euclidean
+              ? std::get<0>(_distance).distanceImpl<int8_t>(x_ptr, y)
+              : std::get<1>(_distance).distanceImpl<int8_t>(x_ptr, y);
+      return distance;
+    }
+
+    throw std::runtime_error("LowPrecisionQuantizer: unsupported bit width");
+  }
+
+  inline size_t getDimension() const { return _dimension; }
+
+  size_t dataSizeImpl() { return _data_size_bytes; }
+
+  void transformDataImpl(void *destination, const void *src) {
+
+    if (_num_bits == 16) {
+      int16_t *quantized_vector = new int16_t[_dimension]();
+      quantizeVector<int16_t>((float *)src, quantized_vector);
+      std::memcpy(destination, quantized_vector, _data_size_bytes);
+
+      delete[] quantized_vector;
+    } else if (_num_bits == 8) {
```

Comment:
can we support sub-char quantization? I think I've seen powers of two for scalar-quant before.